source/host_resolver.c

```diff
@@ -325,11 +366,78 @@ static void s_clear_default_resolver_entry_table(struct default_host_resolver *r
+    (void)status;
+    struct host_purge_callback_options *options = arg;
+    aws_mem_release(options->allocator, task);
+    aws_ref_count_release(&options->ref_count);
+}
+
+static void s_sechdule_purge_cache_callback_async(
+    struct default_host_resolver *default_host_resolver,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+
+    struct host_purge_callback_options *purge_callback_options = s_host_purge_callback_options_new(
+        default_host_resolver->allocator, on_purge_cache_complete_callback, user_data);
+    purge_callback_options->allocator = default_host_resolver->allocator;
+
+    struct aws_task *task = aws_mem_calloc(default_host_resolver->allocator, 1, sizeof(struct aws_task));
+    aws_task_init(task, s_purge_host_cache_callback_task, purge_callback_options, "async_purge_host_address_task");
+
+    struct aws_event_loop *loop = aws_event_loop_group_get_next_loop(default_host_resolver->event_loop_group);
+    AWS_FATAL_ASSERT(loop != NULL);
+    aws_event_loop_schedule_task_now(loop, task);
+}
+
+static int resolver_purge_cache_with_callback(
+    struct aws_host_resolver *resolver,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+
+    if (!on_purge_cache_complete_callback) {
+        return resolver_purge_cache(resolver);
+    }
+
+    struct default_host_resolver *default_host_resolver = resolver->impl;
+    aws_mutex_lock(&default_host_resolver->resolver_lock);
+    struct aws_hash_table *table = &default_host_resolver->host_entry_table;
+
+    /* Success if hash table is already empty. */
+    if (!aws_hash_table_get_entry_count(table)) {
+        aws_mutex_unlock(&default_host_resolver->resolver_lock);
+        s_sechdule_purge_cache_callback_async(default_host_resolver, on_purge_cache_complete_callback, user_data);
+        return AWS_OP_SUCCESS;
+    }
+
+    /* purge all cache */
+    struct host_purge_callback_options *purge_callback_options = purge_callback_options =
+        s_host_purge_callback_options_new(resolver->allocator, on_purge_cache_complete_callback, user_data);
+
+    for (struct aws_hash_iter iter = aws_hash_iter_begin(table); !aws_hash_iter_done(&iter);
+         aws_hash_iter_next(&iter)) {
+        struct host_entry *entry = iter.element.value;
+        /* acquire a refernce to wait for the callback to trigger */
+        aws_ref_count_acquire(&purge_callback_options->ref_count);
+        aws_mutex_lock(&entry->entry_lock);
+        entry->on_host_purge_complete = s_purge_cache_callback;
+        entry->on_host_purge_complete_user_data = purge_callback_options;
+        entry->state = DRS_SHUTTING_DOWN;
+        aws_mutex_unlock(&entry->entry_lock);
+    }
+    aws_hash_table_clear(table);
+
+    /* release the original reference */
+    aws_ref_count_release(&purge_callback_options->ref_count);
```

Comment:
move this outside the lock. the entry will finish purging from another thread, still possible for this release to be the last refcount.